Team mantras (#209)
canonical-web-and-design/practices: team/mantras.md

For example, tables rows may need to stack or become expandible to persist that same information.

## Do not assume a user's intent
Any request for data should be clearly triggered by a user action such as a click or key press. Scrolling or hovering is not an intentional data request from a user.

**Do**: Provide a button or interaction that will clearly request more data from the server.
For example, provide a “Load more…” button at the end of a limited list to request to next page of data.

**Don’t**: On mouseover or scroll request additional data from the server.
For example: a tooltip that requests extra data to populate its contents.

## Reflect state of the application in the URL
All views in an application should be reflected in the URL so a user can share or bookmark an area and jump to it on load. This includes filter and search inputs on listing pages.

**Do**:  When entering an entity details reflect that view in the URL.
For example, /entity/12345

**Don’t**: When clicking to reveal a popover update the address. For example, when filtering and selecting the filters from the drop down do not persist that state in the URL.
For example, /listing/+filter-open

> ## Always allow the user to undo an action

Reviewer: Are these mantras retroactive? If so, this would require significant changes to both MAAS and RBAC (and possibly other apps of ours).

Author: Sure, as guidelines for all patterns and interactions going forward.

Reviewer 2: Yeah, I think it's fine for it to be a principle without us having successfully followed it everywhere. Going forward, we should make every effort to follow it, but there may still be (hopefully extremely rare) exceptions.

Reviewer 3: Undo is awesome (much better than a confirmation step), but sometimes it’s less awesome than other stuff we could implement in the same amount of time. So I think “Always” is too strong here.

Separately, many things can’t or shouldn’t be undoable. For example, formatting or repartitioning a disk, or deleting your PII from a service. One common substitute is providing a timeout. For example, you can’t undo sending e-mail, but in Gmail you get a few seconds to “Undo” (cancel) it instead. You can’t undo deleting your Facebook account, but you get 30 days to change your mind. And often you can’t undo deleting a file/message/photo, but instead items in a trash may expire after a week/month.

Reviewer 2: I still think this is a good principle, and it can lead us to consider implementing interesting solutions like the Gmail delays for non-retractable / descructive actions like formatting.

Agree "always" is probably the wrong word.

---
title: Team mantras
description: A set of high level mantras the web and design team embrace
---

> ## Never remove data or functionality because of screen size

Reviewer 2: For consistency, would it be better for each mantra to be communicated with the same force? As in, if others say "Do not", shouldn't we say "Do not" (or "Don't") here too? Never seems a bit too insistent.

Reviewer 4: I don't think that this is always necessarily the best solution - I know this is going to escalate in a big debate, however I think you should aim to provide the correct amount of information/functionality for the context the user is in. More information/functionality does not always equal a better experience.

Reviewer 2: Fair enough - I suppose I was saying I thought "never" was too strong wording in this case, but I'm open to being persuaded.

Especially, I think the "Never hide" in the next sentence is too strong - because as you suggest, one way to preserve functionality is to hide it behind a user action - click to expand, or a menu or something.

But I take your points. I think they're good.

@lilyvidenova if we expand on the point a bit to clarify that it's okay to *hide* functionality, as long as it's findable, do you still disagree? The point being that nothing you can do or can read on a desktop should be impossible to do / find on a phone?

I'd be interested to hear specific examples where you think it might be justified to remove information or functionality from phone users.

> Never hide information or remove functionally on different screen sizes. In the modern era, applications should be fully featured and easy to use on all devices and screens.

Reviewer 2: ```suggestion
Avoid hiding information or functionally on different screen sizes. In the modern era, applications should be fully featured and easy to use on all devices and screens.
```

> Never hide information or remove functionally on different screen sizes. In the modern era, applications should be fully featured and easy to use on all devices and screens.

Reviewer 3: ```suggestion
Never make information or functions unavailable depending on screen size. Sites should be fully featured and easy to use on all devices and screens.
```

**Do:** If a details page contains a set of metadata and actions which can be performed on that entity. That data and functionality should be accessible on all screens.
For example, when viewing the list on a phone the user should be able to search and filter as they can on a desktop screen.

**Don’t**: Reduce the information on the listing to fit the content in.
For example, tables rows may need to stack or become expandible to persist that same information.

## Do not assume a user's intent

> Any request for data should be clearly triggered by a user action such as a click or key press. Scrolling or hovering is not an intentional data request from a user.

Reviewer 2: I generally agree that we could maybe use slightly less strong language across the board.

Although I think it is interesting that this leads us to question image lazyloading. Another pattern would be to not show images until the user actually clicks to load them (like how flashblock works). In many cases I think that's probably the wrong solution, but it's interesting to be pushed to consider it.

**Do**: Provide a button or interaction that will clearly request more data from the server.
For example, provide a “Load more…” button at the end of a limited list to request to next page of data.

**Don’t**: On mouseover or scroll request additional data from the server.
For example: a tooltip that requests extra data to populate its contents.

> ## Reflect state of the application in the URL

Reviewer 2: ```suggestion
## Reflect the state of the application in the URL
```
